app/src/main/java/com/bignerdranch/android/githublogin/LoginActivity.java

```diff
@@ -0,0 +1,162 @@
+    private static final String AUTH = "auth";
+    private EditText mUsername, mPassword;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        Log.d(TAG, "onCreate() called");
+
+        setContentView(R.layout.activity_main);
+
+        mUsername = (EditText) findViewById(R.id.username);
+        mPassword = (EditText) findViewById(R.id.password);
+        Button loginButton = (Button) findViewById(R.id.login_button);
+        loginButton.setOnClickListener(this);
+
+        // check if already logged in
+        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
+        if (preferences.getString(AUTH, null) != null) {
+           goToProfileScreen();
+        }
+
+        findViewById(R.id.id_container).setOnTouchListener(new View.OnTouchListener(){
+
+            @Override
+            public boolean onTouch(View v, MotionEvent event) {
+                closeKeyboard(v);
+                return false;
+            }
+        });
+
+    }
+
+    @Override
+    public void onClick(View v) {
+        Log.d(TAG, "onClick() called with: v = [" + v + "]");
+        switch (v.getId()) {
+            case R.id.login_button:
+                performLogin();
+                break;
+        }
+    }
+    public void performLogin() {
+        //get credentials
+        final String authHash = Credentials.basic(mUsername.getText().toString(),
+                mPassword.getText().toString());
+
+        Call<LoginData> callable = GitHub.Service.Get()
+                .checkAuth(authHash);
+        callable.enqueue(new Callback<LoginData>() {
+
+            @Override
+            public void onResponse(Call<LoginData> call, Response<LoginData> response) {
+                if(response.isSuccessful()) {
+                    SharedPreferences preferences = PreferenceManager
+                            .getDefaultSharedPreferences(LoginActivity.this);
+                    preferences.edit()
+                            .putString(AUTH, authHash).
+                            putString("username", mUsername.getText().toString())
+                            .apply();
+
+                    goToProfileScreen();
+                } else {
+                    switch(response.code()) {
+                        case 403:
+                            Toast.makeText(LoginActivity.this, "Invalid username or password",
+                                    Toast.LENGTH_SHORT).show();
+                            break;
+                        default:
+                            Toast.makeText(LoginActivity.this, "Failed to login",
+                                    Toast.LENGTH_SHORT).show();
+                            break;
+                    }
+                }
+            }
+
+            @Override
+            public void onFailure(Call<LoginData> call, Throwable t) {
+                t.printStackTrace();
```

Comment:
You can actually check for multiple types of errors here:
1. t is TimeoutException? --> timeout
2. t is UnknownHostException? --> no internet connection